File: Java/src/_0456_132_Pattern/Solution.java
package _0456_132_Pattern;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Date: 2021/3/24 4:02 下午
 * Content: 单调递减栈：从后向前遍历所有元素，使用right记录小于当前考察值的最大值
 * 执行用时：10 ms, 在所有 Java 提交中击败了33.58%的用户
 * 内存消耗：39.4 MB, 在所有 Java 提交中击败了5.01%的用户
 */
public class Solution {

    public boolean find132pattern(int[] nums) {

        Deque<Integer> stk = new ArrayDeque<>();
        int right = Integer.MIN_VALUE;
        for (int i = nums.length - 1; i >= 0; i--) {
            if (nums[i] < right) return true;
            while (!stk.isEmpty() && nums[i] > stk.peek()) {
                right = Math.max(right, stk.remove());
            }
            stk.push(nums[i]);
        }
        return false;
    }

    public static void main(String[] args) {

        // false
        System.out.println((new Solution()).find132pattern(new int[]{1, 2, 3, 4}));
    }
}

File: Python/_0456_132_Pattern/Solution.py
# coding=utf-8
# Date: 2021/6/5 14:44
from typing import List


# 执行用时：188 ms, 在所有 Python3 提交中击败了24.93%的用户
# 内存消耗：26.9 MB, 在所有 Python3 提交中击败了5.05%的用户
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        stk = []
        right = -2e9
        for x in nums[::-1]:
            if x < right:
                return True
            while len(stk) > 0 and x > stk[len(stk) - 1]:
                right = max(right, stk[len(stk) - 1])
                stk.pop(len(stk) - 1)
            stk.append(x)
        return False


if __name__ == "__main__":
    print(Solution().find132pattern([1, 2, 3, 4]))  # False
